src/rust/engine/fs/fs_util/src/main.rs

```diff
@@ -272,65 +319,105 @@ fn save_file(
+      // Exactly one entry with exactly one component indicates either a file in this directory, or
+      // an empty directory.
+      // If the child is a non-empty directory, or a file therein, there must be multiple PathStats
+      // with that prefix component, and we will handle that in the recursive save_directory call.
 
-        let file_futures: Vec<_> = files
-          .into_iter()
-          .map(move |file| {
-            file_to_file_node(store_copy.clone(), posix_fs.clone(), file)
-          })
-          .collect();
+      match path_group.get(0).unwrap() {
+        &fs::PathStat::File { ref stat, .. } => {
+          let stat = stat.clone();
+          file_futures.push(
+            save_file(store.clone(), posix_fs.clone(), stat.clone())
+              .and_then(move |(fingerprint, size_bytes)| {
+                let mut file_node = bazel_protos::remote_execution::FileNode::new();
+                file_node.set_name(osstring_as_utf8(&first_component).unwrap());
+                file_node.set_digest(fingerprint_to_digest(&fingerprint, size_bytes as i64));
+                file_node.set_is_executable(stat.is_executable);
+                Ok(file_node)
+              })
+              .to_boxed(),
+          );
+        }
+        &fs::PathStat::Dir { .. } => {
+          // Because there are no children of this Dir, it must be empty.
+          let (fingerprint, size_bytes) = store
+            .record_directory(&bazel_protos::remote_execution::Directory::new())
+            .unwrap();
+          let mut directory_node = bazel_protos::remote_execution::DirectoryNode::new();
+          directory_node.set_name(osstring_as_utf8(&first_component).unwrap());
+          directory_node.set_digest(fingerprint_to_digest(&fingerprint, size_bytes as i64));
+          dir_futures.push(futures::future::ok(directory_node).to_boxed());
+        }
+      }
+    } else {
+      dir_futures.push(
+        save_directory(
+          store.clone(),
+          posix_fs.clone(),
+          paths_of_child_dir(path_group),
+        ).and_then(move |(fingerprint, size_bytes)| {
+          let mut dir_node = bazel_protos::remote_execution::DirectoryNode::new();
+          dir_node.set_name(osstring_as_utf8(&first_component)?);
+          dir_node.set_digest(fingerprint_to_digest(&fingerprint, size_bytes as i64));
+          Ok(dir_node)
+        })
+          .to_boxed(),
+      );
+    }
+  }
+  join_all(dir_futures)
+    .join(join_all(file_futures))
+    .and_then(move |(dirs, files)| {
+      let mut directory = bazel_protos::remote_execution::Directory::new();
+      directory.set_directories(protobuf::RepeatedField::from_vec(dirs));
+      directory.set_files(protobuf::RepeatedField::from_vec(files));
+      store.record_directory(&directory)
+    })
+    .to_boxed()
+}
 
-        join_all(dir_futures)
-          .join(join_all(file_futures))
-          .to_boxed()
+fn paths_of_child_dir(paths: Vec<fs::PathStat>) -> Vec<fs::PathStat> {
```

Comment:
What I really want to write here is:

`paths.map(|stat| stat { path = path.components().skip(1).collect::<Path>().filter(|p| !p.is_empty()).to_owned() }).collect()`

but:
1. I can't map on a generic "PathStat" without giving PathStat a weird and specific `drop_leading_component` method which doesn't otherwise make any sense, and
2. `components().skip()` returns a `Skip<Component>`, and it's non-obvious how to collect that into a `Path`...

Reply:
> 1. I can't map on a generic "PathStat" without giving PathStat a weird and specific drop_leading_component method

Hm, yea... probably not worth it unless we do it in more than one place.

> 2. components().skip() returns a Skip<Component>, and it's non-obvious how to collect that into a Path...

Yep: that's annoying. But it looks like the following does work (from the examples in the PathBuf docs):
```
fs::PathStat::File {
  path: path.iter().skip(1).collect(),
  stat: stat,
}
```

----

So, code-golfed down (less than you were hoping for), this becomes:
```
paths
  .into_iter()
  .filter(|s| s.path().components().count() == 1)
  .map(|s| match s {
    fs::PathStat::File { path, stat } => {
      fs::PathStat::File {
        path: path.iter().skip(1).collect(),
        stat: stat,
      }
    }
    fs::PathStat::Dir { path, stat } => {
      fs::PathStat::Dir {
        path: path.iter().skip(1).collect(),
        stat: stat,
      }
    }
  })
  .collect()
```

Reply:
Nice! Done, thanks!